### ISSUE TITLE
(eval `(quote ,(fn (x) x))) is broken

### ISSUE THREAD

**Issue by laarc** [sctb/lumen#15]
```
$ bin/lumen
> `(quote ,(fn (x) x))
("quote" function)
> (eval `(quote ,(fn (x) x)))
error: attempt to get length of local 'x' (a function value)
```

`(eval (quote x))` should yield x regardless of the type of x.

I think I'll have time tomorrow to hunt this down.  But if you know where the problem is or how to fix it, feel free.

**Comment by sctb** [comment]
The issue is that a function object is being passed into the compiler (via `eval`), which does not know what to do with it. Indeed, *I* don't know what do with it :)

**Comment by sctb** [comment]
This would be a good test case:
```> ((eval `(quote ,(lambda () 42))))
42```

**Comment by laarc** [comment]
Yeah, the feature turns out to be necessary.  I'll have something up later
today, I think, which will illustrate why.

Intuitively, it might seem like passing a raw function object into eval is
semantically invalid, since there's no way to output compiled code that
contains it.  Indeed, it would be semantically invalid for compile.  But
not for eval.  Eval can treat a raw function object exactly the same as one
that was prepared via compile, as your example illustrates.  And it turns
out that by having that feature, it enables a very interesting runtime
debugger.

**Comment by laarc** [comment]
Implemented in PR #16.

**Comment by laarc** [comment]
"Necessary" is subjective.  Lumen's simplicity probably shouldn't be polluted with this.  It's easy enough for people to add it if they want it.
